Classloader caching

eatkins opened the pull request:
- [x] I've read the [CONTRIBUTING](https://github.com/sbt/sbt/blob/develop/CONTRIBUTING.md) guidelines

This PR unifies and enhances the internal sbt ClassLoaderCache. It generally improves the memory utilization quite a bit. After these changes, I found the metaspace utilization of 1.3.x to be comparable to and sometimes less than 1.2.8. I also found a memory leak that was causing the state to leak with each task evaluation (it was leaking about 3mb of java heap per task evaluation in my test project).

The reason that the metaspace utilization is now in some cases lower is that I was careful to ensure that we always use the state's classLoaderCache when making new loaders. The ClassLoaderCache will evict unused loaders in response to memory pressure. If the MaxMetaspaceSize is _not_ set, we are able to close the underlying URLClassLoader. If that parameter is set, we will leak file handles (which is what sbt < 1.3.0 always did), but the program will not crash if the jvm is able to free up enough space after evicting them. Moreover, there were a number of places where we were making duplicates of the scala-library loader even when it was the same version as the meta build. Re-using the meta build library loader is how the 1.3.x metaspace utilization can be lower than 1.2.8.

I also added a resource layer before the dependency layer. This should allow users to use a dependency layer and still have their code that was dependent on the libraries having access to resource files work. To make this work, I had to remove the ShareRuntimeDependenciesWith... layering strategy. Removing that strategy made it pointless to have RuntimeDependencies and TestDependencies, so I consolidated them into Dependencies. This technically breaks binary compatibility but I think the fact that resources didn't work is a big enough bug that it's worth breaking compatibility.

I also found a path in test where it was possible for the test to throw a NoClassDefError but not report a big warning to the user suggesting that they use the ScalaLibrary or Flat strategies. This should make it less scary to set the default layering strategy to Dependencies, but we still have an opportunity to switch to ScalaLibrary before 1.3.0. FWIW, any build that was working in 1.2.8 should work with ScalaLibrary so I think that is a significantly better option that Flat for the default.

Requires https://github.com/sbt/zinc/pull/674

Fixes #4714, #4711, #4693, #4691, #4686

eatkins committed:
Improve legacy launcher

To minimize classloading and consistency between sbt instances launched
with the latest launcher compared to old launchers, I overhauled code
that replaces the app configuration and meta build classloader at
startup. The goals of this change for legacy launchers were:

1) Do not ever load the scala-library.jar from the app provider class loader.
2) Close the class loaders that are below the topLoader in the class
   loading hierarcy

For the new launcher, we simply want to avoid modifying the loader at
all.

I added the SbtParserInit class so that it was more straightforward to
preload the global instance using reflection. We now use reflection to
instantiate an SbtParserInit instance for both the legacy and new
launcher cases to simplify the logic.

After this change, the legacy loader still uses somewhat more metaspace
than the new loader, but the difference seems to be O(10MB), which
should only impact projects that were close their MaxMetaspaceSize to
begin with.

I verified using javap that none of the code in this class uses the
scala standard library which should help metaspace since we don't load
much of the scala standard library until we enter xMainImpl.run.

eatkins committed:
Add new ClassLoaderCache implementation

This commit adds a new ClassLoaderCache that builds on the
ClassLoaderCache that is present in zinc (and can be used to build an
instance of the zinc ClassLoaderCache to preserve compatibility). It
differs from the zinc classloader cache that it does not use direct
SoftReferences to classloaders. Instead, we create a wrapper loader
that can't load any classes and just delegates to its parent. This
allows us to add a thread that reaps the soft reference to the wrapper
loader. Crucially, we add a custom SoftReference class that has a strong
reference to the underlying classloader. This allows us to call close on
the strong reference.

The one issue with this approach is that we can't
rescue the jvm from crashing with an OOM: metaspace because the jvm
doesn't give us a chance to close and dereference the underlying
classloaders before it crashes. It WILL collect classloaders under
normal memory pressure, just not metaspace pressure. To fix this, I
check if the MaxMetaspaceSize is set via an MxBean and, if it is, we
fill the cache with regular soft references. We are going to change the
bash script to not set -XX:MaxMetaspaceSize by default so most builds
should probably end up correctly closing the classloaders after this
change. But we should break existing builds that set MaxMetaspaceSize
but don't crash.

As part of this commit, I audited all of the places where we were
instantiating ClassLoaderCache instances and instead pass in the
state's ClassLoaderCache instance. This reduces the total number of
classloaders created.

eatkins committed:
Route all ScalaInstance creations through the cache

It was possible to make new classloaders for the scala library and other
jars with each new scala instance. To avoid this, I audited all of the
places within sbt where we make a ScalaInstance and ensure that we
instantiate them in such a way that the classloaders are retrieved
through the state's ClassLoaderCache.

After this change, I found from a heap dump that it was possible to run
test in a project that uses scala 2.12.8 and have only ONE classloader
for the scala library present in the heap dump. With older versions,
there were would be up to 3 or 4 in most heap dumps.

eatkins committed:
Add scripted tests for resources

In sbt 1.3.0-RC1, when the dependency layers are used, the dependencies
cannot see the project resources. This can be problematic when the
library requires certain files like logback.xml to work. This introduces
a scripted test for this case.

The next commit should fix this test.

eatkins committed:
Add scripted watchTriggers

This allows the scripted test files to trigger a continuous scripted
build.

eatkins committed:
Don't use anonymous URLClassLoaders

This makes debugging a bit easier in the eclipse memory analyzer tool
since we get a more specific classloader type than URLClassLoader and by
giving the class a meaningful name, we can tell from where it
originated.

eatkins committed:
Simplify layering strategies

The ShareRuntimeDependenciesLayerWithTestDependencies strategy doesn't
really work with resources, so it makes sense to get rid of it. Without
the share layer, there is no point in having separate
RuntimeDependencies and TestDependencies layers so I consolidated them
to Dependencies.

If we really care about binary/source compatibility for the 1.3.0-RCx
series, I can restore the traits and objects and set them private[sbt].
I think it was kind of a bug that they existed at all given the issue
with resources so it makes sense to just remove them.

eatkins committed:
Don't set fileOutputs at the compile config level

This was problematic because it had no dependency on the compile task
which meant that any other task in the config would pick up those
fileOutputs which did not make sense. I noticed this because
(resources / outputFileStamps).value would include class files.

eatkins committed:
Use file stamps for resource loader

Instead of caching based on the classpath of the resources, we should
instead cache based on the actual resource files. This commit achieves
that by adding the classpathFiles key which just transforms the
attributed classpath to a Seq[Path]. This implicitly generates the
outputFileStamps key for classpathFiles which we can use to read the
stamps (the file stamp entries for the classpath should get filled by
the compile task so this shouldn't actually cause any additional io).

eatkins committed:
Use new ClassLoaderCache for layered classloaders

This commit removes the ClassLoaderCache that I'd added for the purpose
of caching layered classloaders. Instead, we will use the state's global
ClassLoaderCache. This is better both because it centralizes the
classloader caching and because the new ClassLoaderCache will evict
unused classloaders when the jvm is under memory pressure.

I also add a new layer for the resources that goes between the scala
library layer and the dependency layer. This should help in cases where
users depend on libraries that require access to resources, e.g.
logback.xml.

eatkins committed:
Catch NoClassDefFoundErrors in tests

If a test threw a NoClassDefFoundError, it was uncaught and the user got
an obscure error. After this change, they will be warned that the
classloader layering strategy may be at fault and will get instructions
on how to fix it.

eatkins committed:
Fix memory leak

Using a lazy val causes the log manager to hold onto a reference to the
state. These would accumulate with each task evaluation. I found that
that in the beanpuree project, that if I ran compile 10 times in a row,
the heap usage was 40mb lower after this change.

eatkins committed:
Fix dotty plugin

The dotty sbt-bridge module assumes that it's going to get a
URLClassLoader from which it can extract all of the classpath urls. That
doesn't work with the old wrapped classloader because its classpath was
empty. As a nasty workaround, I override the getURLs method, which is
where it gets the URLs from. After this change the
compiler-project/dotty-compiler-plugin test passes.

eed3si9n commented on main/src/main/scala/sbt/ClassLoaderLayeringStrategy.scala:
```diff
@@ -89,36 +89,12 @@ object ClassLoaderLayeringStrategy {
+   * Add a layer on top of the ScalaLibrary layer for all of the task jar dependencies.
    */
-  sealed trait RuntimeDependencies extends ScalaLibrary
+  sealed trait Dependencies extends ScalaLibrary
```
The term "dependencies" comes up too often in the context of sbt, including `import Dependencies._`, so I think `RuntimeDependencies` is a bit more clearer than `Dependencies` even if it means sometimes test dependencies too.

eatkins replied:
I'm open to a completely different name then.

eatkins replied:
Kinda ugly, but maybe `ScalaLibraryThenDependencyJars?` It's at least fairly clear.

eed3si9n replied:
`LibraryJars`, `ExternalJars`, `LayerDependencyJars`?

eatkins replied:
I think of those I like ExternalJars the best just because ScalaLibrary already has Library in it.

eatkins replied:
I picked `AllLibraryJars` to distinguish from `ScalaLibrary`.

eed3si9n commented on main-command/src/main/scala/sbt/internal/classpath/ClassLoaderCache.scala:
```diff
@@ -0,0 +1,217 @@
+private object ClassLoaderCache {
+  private def threadID = new AtomicInteger(0)
+}
+private[sbt] class ClassLoaderCache(val configuration: AppConfiguration)
```
It might be good to add some unit tests for this new cache to test some of the invariants?

eatkins replied:
Oh, I actually have a bunch of tests of the old ClassLoaderCache in the main project that I forgot to delete and won't compile now anyway. I'll update the PR tomorrow with the api changes needed to make them run in the command project.

eatkins replied:
https://github.com/sbt/sbt/pull/4729/commits/53d609669cef1a1b0af1e4eb94b267a5d3538003 ports the old ClassLoaderCacheTest. The new cache doesn't support a size limit (yet) so I removed one of the tests. It isn't clear to me that there is much benefit to manually specifying the size limit since the actual size impact of each classloader can vary greatly and we can't easily measure it. For now, I think just having the jvm manage the cache implicitly through the soft references seems like the best approach until we are sure that it's a problem.

eed3si9n commented on main-command/src/main/scala/sbt/internal/classpath/ClassLoaderCache.scala:
```diff
@@ -0,0 +1,217 @@
+    }
+  }
+  private[this] val mkReference: (Key, ClassLoader) => Reference[ClassLoader] =
+    if (metaspaceIsLimited)(_, cl) => new SoftReference(cl, referenceQueue)
```
I can't think of other places in sbt where we are doing something special based on the memory parameter like this.
It might be good to leave some comment here explaining the current tradeoff.

eatkins replied:
Fair enough. It is mentioned in the commit message but this probably deserves higher visibility.

eatkins replied:
https://github.com/sbt/sbt/pull/4729/commits/61bfba0dedd77fa12d7deebc6e6b69c21f18219f

eed3si9n commented on main/src/main/scala/sbt/Defaults.scala:
```diff
@@ -717,8 +717,19 @@ object Defaults extends BuildCommon {
+          Def.task {
+            try {
+              val allJars = scalaProvider.jars
+              val Array(libraryJar) = allJars.filter(_.getName == "scala-library.jar")
```
1. I don't think we can assume that the name `scala-library.jar` will remain that way for all implementations of Scala. That's why ScalaInstance returns `libraryJar`.
2. As of https://github.com/sbt/zinc/pull/677, it can return multiple files as `libraryJars`.

eatkins replied:
Yeah. That is why I catch the exception and revert to the old approach if that doesn’t work (though I didn’t mean to leave the e.printStackTrace() in). I can look into incorporating the change from https://github.com/sbt/zinc/pull/677 though since I require a later version if zinc anyway.

eatkins replied:
I figure though that any future changes to the scala compiler api require zinc/sbt coordinations anyway so I don’t know that it’s really possible to future proof this.

eatkins replied:
I reworked this commit to use the new libraryJars api. The specific code block where you were worried about assuming that the scala library is called "scala-library.jar" seems completely safe the more I think about it. We only execute that code if the scalaVersion is the same as the application scala provider version. I can't figure out how that assumption would be broken without a lot of other things breaking. We also already assume in numerous places that the jars have speciifc names, see e.g. sbt.librarymanagement.ScalaArtifacts. I did, however, rework that code block to be more obviously safe in https://github.com/sbt/sbt/pull/4729/commits/d05344527ee670c6a07be28a3e48626235b9cab8.

eatkins committed:
Add comment explaining metaspace argument behavior

eatkins committed:
Bump launcher

This new versions layers the classpath so that the
test-interface-1.0.jar appears before the scala library. This allows us
to bypass reconstructing the AppConfiguration at startup which reduces
the metaspace utilization of sbt.

eatkins committed:
Don't automatically die on OOM: metaspace

In an interactive session, it's possible for task evaluation to trigger
an OOM: Metaspace but for sbt to continue working after that failure.
Moreover, the metaspace oom can be caused by using a dependency
classloader layer. If the user changes the layering strategy, they may
be able to re-run their command successfully.

eatkins committed:
Bump zinc dependency

This supports a custom ClassLoaderCache and multiple scala library jars
in the ScalaInstance.

eatkins committed:
Don't set sbt version in .travis.yml

We should test the version specified in build.properties.

eatkins committed:
Add missing header

eatkins committed:
Remove dead test

eatkins committed:
Use libraryJars rather than libraryJar in ClassLoaders

Dotty uses multiple library jars. It also simplifies the code to use the
libraryJars method.

eatkins committed:
Set mainProj/Test/classLoaderLayeringStrategy

The TestDependencies strategy doesn't work as of 1.3.0-RC1 because some
of the tests need to access resources. The ScalaLibrary strategy is
broken in 1.3.0-RC1. That leaves Flat. I tested that with 1.3.0-SNAPSHOT
off the latest master, that the AllLibraryJars strategy works fine, but
I need this change now to get travis to work.

eatkins committed:
Don't check parents in ClassLoaderCacheTest

This check doesn't actually make sense anymore with the new
ClassLoaderCache. In the old ClassLoaderCache, there were separate
layers for the snapshots and regular jars.  The test was verifying that
only the snapshot layer was invalidated but now there is just one layer.

eatkins commented:
Finally green on CI.
